system/jlib/jthread.cpp

```diff
@@ -1954,15 +1954,14 @@ protected: friend class PipeWriterThread;
         int inpipe[2];
         int outpipe[2];
         int errpipe[2];
-        if (hasinput)
-            if (::pipe(inpipe)==-1)
-                throw makeOsException(errno);
-        if (hasoutput)
-            if (::pipe(outpipe)==-1)
-                throw makeOsException(errno);
-        if (haserror)
-            if (::pipe(errpipe)==-1)
-                throw makeOsException(errno);
+        if ((hasinput && (::pipe(inpipe)==-1)) ||
+            (hasoutput && (::pipe(outpipe)==-1)) ||
+            (haserror && (::pipe(errpipe)==-1)))
+        {
+            retcode = START_FAILURE;
+            started.signal();
+            throw makeOsException(errno);
```

Comment:
Just a small comment, probably could/should throw(errno) below if fork() fails to be consistent.

Reply:
I agree fork() failing should also be a makeOsException.
But not new, use to throw here, use not to throw there - nothing changed  as far as that is concerned. 
This is fixing the deadlock.

Could open another JIRA to address that though?